src/AEG-IO.ts

```diff
@@ -48,35 +54,62 @@ export async function saveFile(handle: FileSystemFileHandle, aegData: AEGTree) {
 
 /**
  * Function that takes in data read from a file and converts it into a valid AEG representation.
+ * @param mode The mode we are in (Draw mode or proof mode)
  * @param fileData The data read from a file.
- * @returns An AEG representation of the data.
+ * @returns If in draw mode, returns an AEG representation of the data.
+ * If in proof mode, constructs an array of AEGs read from the file.
+ * This can be used to build the proof list
  * Returns null if an error occurred
  */
-export function loadFile(fileData: string | ArrayBuffer | null): AEGTree | null {
+export function loadFile(
+    mode: "Draw" | "Proof",
+    fileData: string | ArrayBuffer | null
+): AEGTree | ProofNode[] | null {
     if (typeof fileData === "string") {
-        const data: sheetObj = JSON.parse(fileData);
-        const childData: (atomObj | cutObj)[] = data.internalSheet.internalChildren;
-
-        const tree: AEGTree = new AEGTree();
-        const children: (AtomNode | CutNode)[] = [];
-
-        childData.forEach(child => {
-            if (Object.prototype.hasOwnProperty.call(child, "internalEllipse")) {
-                //make cut
-                children.push(toCut(child as cutObj));
-            } else {
-                //Make atom
-                children.push(toAtom(child as atomObj));
-            }
-        });
-
-        tree.sheet.children = children;
-        return tree;
+        const data = JSON.parse(fileData);
+
+        if (mode === "Draw") {
+            const childData: (atomObj | cutObj)[] = (data as sheetObj).internalSheet
+                .internalChildren;
+            return toTree(childData);
+        } else {
+            //Construct the tree at every step of the proof and store them in an array
+            const arr: ProofNode[] = [];
+            data.forEach((node: nodeObj) => {
+                const childData: (atomObj | cutObj)[] = node.tree.internalSheet.internalChildren;
+                arr.push(new ProofNode(toTree(childData), node.appliedRule));
+            });
+
+            return arr;
+        }
     }
 
     return null;
 }
 
+/**
+ * Constructs an AEG from the array of JSON objects parsed from our file data.
+ * @param childData The array of objects which should be filled in as children of the tree.
+ * @returns An AEG Tree representation of our data.
+ */
+function toTree(childData: (atomObj | cutObj)[]): AEGTree {
+    const tree: AEGTree = new AEGTree();
+    const children: (AtomNode | CutNode)[] = [];
+
+    childData.forEach(child => {
```

Comment:
Avoid .forEach when you could use a for-of loop instead. (For of loops are ~25% faster than .forEach)

See https://stackoverflow.com/questions/50844095/should-one-use-for-of-or-foreach-when-iterating-through-an-array